### PR TITLE
Added info extractor for www.tele5.de, a German TV station.

### PR DESCRIPTION
**Pull request by jensrutschmann**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information
This adds support for www.tele5.de, which is a German TV station providing temporary online access to movies and TV shows after broadcast.

I have added tests, but since the referenced videos will only be available for limited time, these tests will fail in the future. They all were successful at the time this PR has been created, though. Unfortunately the site does not provide videos that are available for a longer time.

### CODE REVIEW

File: youtube_dl/extractor/tele5.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+from .nexx import NexxIE
+
+
+class Tele5IE(InfoExtractor):
+    _VALID_URL = r'https://www.tele5.de/(mediathek/filme-online/videos|tv/).*'
```
Comment (dstftw) [inline youtube_dl/extractor/tele5.py]: 1. `.*` at the end is senseless.
2. Don't capture groups you don't use.
3. Capture and use display id.

File: youtube_dl/extractor/tele5.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        }
+    }, {
+        'url': 'https://www.tele5.de/tv/buffy-im-bann-der-daemonen/videos',
+        'info_dict': {
+            'id': '1547129',
+            'ext': 'mp4',
+            'upload_date': '20180730',
+            'timestamp': 1532967491,
+            'title': 'Der Höllenhund'
+        }
```
Comment (dstftw) [inline youtube_dl/extractor/tele5.py]: Remove all duplicate tests.

File: youtube_dl/extractor/tele5.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    }]
+
+    def _real_extract(self, url):
+        webpage = self._download_webpage(url, 'N/A')
+
+        id = self._html_search_regex(
+            r'class="ce_videoelementnexx-video__player"\sid="video-player"\sdata-id="(?P<id>[0-9]+)"',
```
Comment (dstftw) [inline youtube_dl/extractor/tele5.py]: Should be more relaxed.

File: youtube_dl/extractor/tele5.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            'upload_date': '20180730',
+            'timestamp': 1532967491,
+            'title': 'Der Höllenhund'
+        }
+    }]
+
+    def _real_extract(self, url):
+        webpage = self._download_webpage(url, 'N/A')
+
+        id = self._html_search_regex(
```
Comment (dstftw) [inline youtube_dl/extractor/tele5.py]: Don't shadow built-in names.

File: youtube_dl/extractor/tele5.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+            r'class="ce_videoelementnexx-video__player"\sid="video-player"\sdata-id="(?P<id>[0-9]+)"',
+            webpage, 'id')
+
+        return self.url_result(
+            'https://api.nexx.cloud/v3/759/videos/byid/%s'
+            % id, ie=NexxIE.ie_key())
```
Comment (dstftw) [inline youtube_dl/extractor/tele5.py]: Pass video id.

**Comment by jensrutschmann** [comment]
@dstftw I've implemented the requested changes - sorry for the beginner bugs - I've never done python programming before.

The only requested change I've not implemented is 'Capture and use display id.' because not all URLs contain the actual video ID.
For example the URL from the second test always points to the current (and only available) episode of a series, so the actual video reachable via this URL will change - for example each week, when a new episode is made available.

**Comment by dstftw** [comment]
I'm not talking about video id, capture part of URL as display id.

File: youtube_dl/extractor/tele5.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+from .nexx import NexxIE
+
+
+class Tele5IE(InfoExtractor):
+    _VALID_URL = r'https://www.tele5.de/[mediathek/filme-online/videos|tv/]'
```
Comment (dstftw) [inline youtube_dl/extractor/tele5.py]: Incorrect.

File: youtube_dl/extractor/tele5.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        }
+    }, {
+        'url': 'https://www.tele5.de/tv/dark-matter/videos',
+        'info_dict': {
+            'id': '1548206',
+            'ext': 'mp4',
+            'title': 'Folge Sechsundzwanzig',
+            'timestamp': 1533664358,
+            'upload_date': '20180807'
+        }
```
Comment (dstftw) [inline youtube_dl/extractor/tele5.py]: I've already pointed out: **remove all duplicate tests**. Or make them only_matching.

File: youtube_dl/extractor/tele5.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    }]
+
+    def _real_extract(self, url):
+        webpage = self._download_webpage(url, 'N/A')
+
+        video_id = self._html_search_regex(
+            r'id="video-player"\sdata-id="(?P<id>[0-9]+)"',
```
Comment (dstftw) [inline youtube_dl/extractor/tele5.py]: Must allow arbitrary whitespace not only `\s`. Must allow single quotes. Group name is senseless when there is only one group.

File: youtube_dl/extractor/tele5.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+            webpage, 'id')
+
+        return self.url_result(
+            'https://api.nexx.cloud/v3/759/videos/byid/%s'
+            % video_id, ie=NexxIE.ie_key())
```
Comment (dstftw) [inline youtube_dl/extractor/tele5.py]: My comment here was ignored.

File: youtube_dl/extractor/tele5.py (hunk 7 above)
Comment (jensrutschmann) [reply youtube_dl/extractor/tele5.py]: Why is this a duplicate test although it tests another URL pattern?

Comment (dstftw) [reply youtube_dl/extractor/tele5.py]: Because all of them use the same extraction scenario.

File: youtube_dl/extractor/tele5.py (hunk 6 above)
Comment (dstftw) [reply youtube_dl/extractor/tele5.py]: Read some doc on regular expressions. Namely what does `[]` and `()` mean and how it should be used.

**Comment by jensrutschmann** [comment]
@dstftw I think I've implemented all requested changes.
Let me know if there's anything missing or if I should squash the commits before merging.

**Comment by mrtnmtth** [comment]
Here is my take on it:  
https://github.com/mrtnmtth/youtube-dl/blob/master/youtube_dl/extractor/tele5.py

My regex is more relaxed by making `www` optional and allowing `http` without `s`.  
Also I only download parse the html if there is no id in the url.

Feel free to copy parts of it as you like.

File: youtube_dl/extractor/tele5.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+            'title': 'SchleFaZ: Atomic Shark',
+        }
+    }, {
+        'url': 'https://www.tele5.de/tv/dark-matter/videos',
+        'only_matching': True,
+    }]
+
+    def _real_extract(self, url):
+        mobj = re.match(self._VALID_URL, url)
+        display_id = mobj.group('display_id')
```
Comment (dstftw) [inline youtube_dl/extractor/tele5.py]: `self._match_id`.

File: youtube_dl/extractor/tele5.py [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+        webpage = self._download_webpage(url, display_id)
+
+        video_id = self._html_search_regex(
```
Comment (dstftw) [inline youtube_dl/extractor/tele5.py]: `_search_regex`.

File: youtube_dl/extractor/tele5.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+from __future__ import unicode_literals
+
+import re
+
+from .common import InfoExtractor
+from .nexx import NexxIE
+
+
+class Tele5IE(InfoExtractor):
+    _VALID_URL = r'https://www\.tele5\.de/(?:mediathek/filme-online/videos\?vid=|tv/)(?P<display_id>[\w-]+)'
```
[the remark this answers is not in the document]
Comment (mrtnmtth) [reply youtube_dl/extractor/tele5.py]: My take is:  
`_VALID_URL = r'https?://(?:www\.)?tele5\.de/(?:mediathek/filme-online/videos(/|\?(.*&)?vid=)|tv/)(?P<id>[\w-]+)'`

It matches all of the following:
```
https://www.tele5.de/mediathek/filme-online/videos?blah=1&vid=1550589
https://www.tele5.de/mediathek/filme-online/videos/schlefaz-atomic-shark
https://www.tele5.de/mediathek/filme-online/videos?vid=1549415
https://www.tele5.de/tv/digimon/videos
https://www.tele5.de/tv/digimon/videos/rikas-krise
https://www.tele5.de/tv/kalkofes-mattscheibe/video-clips/politik-und-gesellschaft?ve_id=1551191
```

**Pull request closed without merging**